Rest Client: Add property to skip hostname verification

Sgitario opened the pull request:
Before these changes, we only can disable the hostname verification in Rest Client classic by providing the following property:

```
quarkus.rest-client.extensions-api.hostname-verifier=io.quarkus.restclient.NoopHostnameVerifier
```

However, this is not working in Rest Client reactive because setting a hostname verifier strategy is not supported by the Vert-x HTTP Client. 

With these changes, we have added a new property in both Rest Client classic and reactive `quarkus.rest-client.extensions-api.verify-host=true or false`. In Rest Client classic, when disabling the verify host, internally it will add the `NoopHostnameVerifier` strategy. In Rest Client reactive, it will properly configure the Vert.x HTTP client to disable the hostname verification. Therefore, in both Rest Client implementations (classic and reactive), the behaviour is the same. 

Fix https://github.com/quarkusio/quarkus/issues/27901

geoand commented:
@famod mind trying this out?

famod commented:
@geoand I'd love to, but right now I'm standing in the middle of chaos because on the weekend we're finally relocating to our new house.
So better not wait for me.

geoand commented:
Hope all goes well with the move :)

geoand commented:
@Chexpir would you like to test this before we merge it?

gsmet reviewed (dismissed):
Asked an important question :).

gsmet commented on docs/src/main/asciidoc/rest-client-reactive.adoc:
```diff
@@ -238,6 +238,15 @@ quarkus.rest-client.extensions-api.url=https://stage.code.quarkus.io/api
+
+[source,properties]
+----
+quarkus.rest-client.extensions-api.verify-host=false
```
Asking that because somebody asked with the behavior of `quarkus.tls.trust-all` in conjunction with REST Clients. That being said, being able to adjust things specifically for REST Client would be useful anyway so I think this patch has value.

cescoffier replied:
Yes, `quarkus.tls.trust-all` is global. Having a rest client attribute would be nice. However, I would keep the name `trust-all`.

famod replied:
Are you talking about a (per-client) `trust-all` that skips more verifications _instead_ of `verify- host` that skips only the hostname verification?
Or is this just a naming discussion?

cescoffier replied:
Trust-all by-pass everything (basically pass a ssl context that accept any certs).

famod replied:
Yes, that's why I'm asking. `trust-all` (global or per client) is a sledgehammer which is usually unacceptable in production (where you might have cases though in which you can only use an IP to talk to another system).
So my point is that `trust-all` (on whichever level) should not "replace" `verify-host`.

So, on client level, `trust-all` _additionally_ to `verify-host` would be ok for me, but having _only_ `trust-all` is much too coarse.

geoand replied:
I agree with @famod on this.

Chexpir commented:
@geoand thanks for the proposition, no need to test it. I've reviewed the code and I trust that it will work!
Anyway, it will be useful to know the process to build and test github branches, it might become handy in the future, as we are into Quarkus for the long term! (I have searched and could not find anything).

geoand commented:
I would say that the easiest way to checkout a Pull Request is following [this](https://cli.github.com/manual/gh_pr_checkout).
Once that is done, you can build it locally by following [this](https://github.com/quarkusio/quarkus/blob/main/CONTRIBUTING.md#build).

geoand commented:
@Chexpir were you able to test this one?

geoand commented:
Were are we with this one?

I personally think we should get it in

Sgitario commented:
> Were are we with this one?
> 
> I personally think we should get it in

My +100 to get it in.

cescoffier commented:
My last comment was to make sure we clearly say in the doc that skipping host name verification should not be used in prod ('use it to your own risk')

geoand commented:
We should 100% add that

Sgitario committed:
Rest Client: Add property to skip hostname verification

Before these changes, we only can disable the hostname verification in Rest Client classic by providing the following property:

```
quarkus.rest-client.extensions-api.hostname-verifier=io.quarkus.restclient.NoopHostnameVerifier
```

However, this is not working in Rest Client reactive because setting a hostname verifier strategy is not supported by the Vert-x HTTP Client. 

With these changes, we have added a new property in both Rest Client classic and reactive `quarkus.rest-client.extensions-api.verify-host=true or false`. In Rest Client classic, when disabling the verify host, internally it will add the `NoopHostnameVerifier` strategy. In Rest Client reactive, it will properly configure the Vert.x HTTP client to disable the hostname verification. Therefore, in both Rest Client implementations (classic and reactive), the behaviour is the same. 

Fix https://github.com/quarkusio/quarkus/issues/27901

Sgitario commented:
PR updated with adding this warning to the docs.

Sgitario commented:
Any updates on this pull request? I would like to proceed to merge it if everybody agrees.

fasuke85 commented:
Hi, it would be awsome if you merged this PR 😄

oven commented:
This issue is a blocker for us because we have to communicate with an HTTPS service over an ssh tunnel. Please merge :)

geoand commented:
@gsmet any last words or should I dismiss the old review?

geoand commented:
Since the last CI run of this was a while ago, I rebased onto `main`.
